Feature/db connect

pullveryzator opened the pull request:
Написал crud для взаимодействия с базой. По сути сделал их общими для всех моделей. Любая критика и предложения приветствуются. Наверняка что-то упустил...

pullveryzator committed:
Write base CRUD operations.

On branch feature/db_connect
Changes to be committed:
	new file:   src/app/core/__init__.py
	new file:   src/app/core/db.py
	new file:   src/app/crud/__init__.py
	new file:   src/app/crud/base.py
	new file:   src/app/crud/group.py
	new file:   src/app/crud/message.py
	new file:   src/app/crud/photo.py
	new file:   src/app/crud/user_tg.py
	new file:   src/app/models/__init__.py
	new file:   src/app/schemas/__init__.py
	modified:   src/bot/core/settings.py

pullveryzator committed:
Small fix function name.

On branch feature/db_connect
Changes to be committed:
	modified:   src/app/core/db.py
	modified:   src/app/crud/base.py

pullveryzator committed:
Style fix

Changes to be committed:
	modified:   src/app/crud/base.py

pullveryzator committed:
Поправил импорты в crud c учетом актуальных имен моделей.
Changes to be committed:
	modified:   src/app/core/db.py
	modified:   src/app/crud/group.py
	modified:   src/app/crud/message.py
	modified:   src/app/crud/photo.py
	modified:   src/app/crud/user_tg.py

teamofroman commented on src/app/crud/user_tg.py:
```diff
@@ -0,0 +1,11 @@
+from app.models.models import UserTG
+
+
+class CRUDUserTG(CRUDBase):
```
Для пользователя надо переопределить метод `create` и `update`, т.к. в схеме будет поле `password` с паролем, а в БД мы храним его hash. Т.о. надо будет с словаре, который получится из схемы заменить `password` на `hashed_password` и вписать туда значение хешированного пароля.

pullveryzator replied:
С этим пока думаю..

pullveryzator replied:
Реализовал на встроенной функции hash().

teamofroman commented on src/app/crud/base.py:
```diff
@@ -0,0 +1,93 @@
+        """Update object in database."""
+        obj_data = jsonable_encoder(db_obj)
+        update_data = pydantic_scheme_obj.dict(exclude_unset=True)
+        for field in obj_data:
```
Давай пойдем ноборот. Идем по `update_data`, проверяем, есть ли у `db_obj` аттрибут `field` (`hasattr`) и если есть, устанавливаем его значение.

pullveryzator replied:
Поправил.

teamofroman commented on src/app/crud/base.py:
```diff
@@ -0,0 +1,93 @@
+            session: AsyncSession) -> ModelType:
+        """Update object in database."""
+        obj_data = jsonable_encoder(db_obj)
+        update_data = pydantic_scheme_obj.dict(exclude_unset=True)
```
Давай еще добавим исключение значений, которые установлены в None.

pullveryzator replied:
Поправил.

teamofroman commented on src/bot/core/settings.py:
```diff
@@ -19,6 +19,13 @@ class Settings(BaseSettings):
 
     email_curator: str = ''
 
+    postgres_user = ''
```
Давай для единообразия кода добавим аннатации.
Значения по умолчанию убираем. Без этих данных мы не подключимся к БД.

pullveryzator replied:
Поправил. Вопрос: нужно ли убирать значения по-умолчанию у других параметров? (в legacy)

teamofroman commented on src/bot/core/settings.py:
```diff
@@ -19,6 +19,13 @@ class Settings(BaseSettings):
+    postgres_port = ''
+    database_url = ''
+
     class Config:  # noqa: D106
```
Давай добавим `extra = 'ignore'`, что бы не валиться на тех настройках, которые есть в env-файле, но нет в классе.

pullveryzator replied:
Поправил.

teamofroman reviewed (changes requested):
Отлично.
Есть несколько моментов, которые надо поправить.

teamofroman commented on src/app/crud/base.py:
```diff
@@ -0,0 +1,93 @@
+        """Initilisation method for CRUDBase class."""
+        self.model = model
+
+    async def _get_by_attribute(
```
Давай сделаем так, что бы можно было делать фильтрацию по нескольким атрибутам. Все их соединяем по условию `И`

pullveryzator replied:
Сделал по нескольким параметрам. Условия должны приходить в виде словаря.

pullveryzator committed:
Fix crud. Add check_tg_id_unique courutine.

Changes to be committed:
	modified:   src/app/crud/base.py
	modified:   src/app/crud/user_tg.py
	new file:   src/app/endpoints/__init__.py
	new file:   src/app/endpoints/validators.py
	modified:   src/bot/core/settings.py

teamofroman commented on src/app/endpoints/validators.py:
```diff
@@ -0,0 +1,21 @@
+) -> None:
+    """Check unique telegram id for new user."""
+    tg_id = pydantic_scheme_user.tg_id
+    existing_user_query = select(UserTG).where(UserTG.tg_id == tg_id)
```
Для проверки существования записи лучше воспользоваться функцией `exists`, т.к. сам объект нам не нужен.
https://dnmtechs.com/using-sqlalchemys-existence-query-in-python-3/

pullveryzator replied:
Поправил.

teamofroman commented on src/app/crud/user_tg.py:
```diff
@@ -0,0 +1,55 @@
+            session: AsyncSession) -> ModelType:
+        """Create new user in database."""
+        new_user_dict = pydantic_scheme_user.dict()
+        await check_tg_id_unique(pydantic_scheme_user, session)
```
Проверка существования пользователя не задача функции `create`. Ее задача - создать запись в базе. Убеждаться что пользователя с таким tg_id нет должна вызывающая функция.

Аналогично для `update`

pullveryzator replied:
Поправил.

teamofroman commented on src/app/crud/user_tg.py:
```diff
@@ -0,0 +1,55 @@
+        new_user_dict = pydantic_scheme_user.dict()
+        await check_tg_id_unique(pydantic_scheme_user, session)
+        password = new_user_dict.pop('password')
+        new_user_dict['hashed_password'] = hash(password)
```
Здесь надо добавить комментарий TODO с пометкой "заменить на получение хеша после создания функций для авторизации"

`# TODO: ...`

Аналогично для `update`

pullveryzator replied:
Поправил.

teamofroman commented on src/app/endpoints/validators.py:
```diff
@@ -0,0 +1,21 @@
+    result = await session.execute(existing_user_query)
+    existing_user = result.scalars().first()
+
+    if existing_user:
```
Выбрасывать исключение здесь не надо. Достаточно вернуть `True`, если пользователя с таким tg_id нет, и `False` в противном случае.

И т.к. эта функция относится к пользователям, ее можно смело включить в CRUD для работы с пользователем.

pullveryzator replied:
Поправил.

teamofroman reviewed (changes requested):
Есть небольшие правки

pullveryzator committed:
fix user_tg crud.
Changes to be committed:
	modified:   src/app/crud/user_tg.py
	deleted:    src/app/endpoints/validators.py

teamofroman committed:
Update user_tg.py

teamofroman reviewed (approved):
Принято

teamofroman committed:
Merge branch 'develop' into feature/db_connect

teamofroman committed:
Update user_tg.py

teamofroman committed:
Update user_tg.py

teamofroman committed:
Style fix
